exporter/prometheusremotewriteexporter/wal.go

```diff
@@ -0,0 +1,192 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package prometheusremotewriteexporter
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"path/filepath"
+	"sync"
+	"time"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/prometheus/prometheus/prompb"
+	"github.com/tidwall/wal"
+)
+
+type prweWAL struct {
+	mu        sync.Mutex // mu protects the fields below.
+	wal       *wal.Log
+	walConfig *walConfig
+	walPath   string
+
+	exportSink func(ctx context.Context, reqL []*prompb.WriteRequest) []error
+
+	stopOnce  sync.Once
+	stopChan  chan struct{}
+	rWALIndex uint64
+	wWALIndex uint64
+}
+
+func newWAL(walConfig *walConfig, exportSink func(context.Context, []*prompb.WriteRequest) []error) (*prweWAL, error) {
+	if walConfig == nil {
+		// There are cases for which the WAL can be disabled.
+		// TODO: Perhaps log that the WAL wasn't enabled.
+		return nil, errNilConfig
+	}
+
+	return &prweWAL{
+		exportSink: exportSink,
+		walConfig:  walConfig,
+		stopChan:   make(chan struct{}),
+	}, nil
+}
+
+func (wc *walConfig) createWAL() (*wal.Log, string, error) {
+	walPath := filepath.Join(wc.Directory, "prom_remotewrite")
+	wal, err := wal.Open(walPath, &wal.Options{
+		SegmentCacheSize: wc.nBeforeTruncation(),
+		NoCopy:           true,
+	})
+	if err != nil {
+		return nil, "", fmt.Errorf("prometheusremotewriteexporter: failed to open WAL: %w", err)
+	}
+	return wal, walPath, nil
+}
+
+type walConfig struct {
+	// Note: These variable names are meant to closely mirror what Prometheus' WAL uses for field names per
+	// https://docs.google.com/document/d/1cCcoFgjDFwU2n823tKuMvrIhzHty4UDyn0IcfUHiyyI/edit#heading=h.mlf37ibqjgov
+	// but also we are using underscores "_" instead of dashes "-".
+	Directory         string        `mapstructure:"directory"`
+	NBeforeTruncation int           `mapstructure:"n_before_truncation"`
+	TruncateFrequency time.Duration `mapstructure:"truncate_frequency"`
+}
+
+func (wc *walConfig) nBeforeTruncation() int {
+	if wc.NBeforeTruncation <= 0 {
```

Comment:
Should this also enforce an upper bound?

Reply:
Not really, would be arbitrary to enforce one given that for high throughput apps we could even want say 50,000 items first before truncation or for some 100.